{"cells":[{"metadata":{},"cell_type":"markdown","source":"<h2 align='middle'>OpenVaccine: COVID-19 mRNA Vaccine Degradation Prediction</h2>\n<h3 align='middle'>Urgent need to bring the COVID-19 vaccine to mass production</h3>\n<div align='middle'> \n    <img src='https://media.giphy.com/media/dVuyBgq2z5gVBkFtDc/giphy.gif' alt='COIVD-19 (GIPHY)'></img>\n</div>"},{"metadata":{},"cell_type":"markdown","source":"### What is RNA?\n\n**Ribonucleic acid**, or RNA is one of the three major biological macromolecules that are essential for all known forms of life (along with DNA and proteins). A central tenet of molecular biology states that the flow of genetic information in a cell is from DNA through RNA to proteins: **“DNA makes RNA makes protein”**.\n\n---\n\n### What is mRNA?\n\n**Messenger RNAs**, also known as mRNA, are one of the types of RNA that are found in the cell. This particular one, like most RNAs, are made in the nucleus and then exported to the cytoplasm where the translation machinery, the machinery that actually makes proteins, binds to these mRNA molecules and reads the code on the mRNA to make a specific protein. So in general, one gene, the DNA for one gene, can be transcribed into an mRNA molecule that will end up making one specific protein.\n\n---\n\n### Why are we here?\n\nmRNA vaccines have taken the lead as the fastest vaccine candidates for COVID-19, but currently, they face key potential limitations. One of the biggest challenges right now is how to design super stable messenger RNA molecules (mRNA). Conventional vaccines (like your seasonal flu shots) are packaged in disposable syringes and shipped under refrigeration around the world, but that is not currently possible for mRNA vaccines.\n\nResearchers have observed that RNA molecules have a tendency to spontaneously degrade. This is a serious limitation--a single cut can render the mRNA vaccine useless. Currently, little is known on the details of where in the backbone of a given RNA is most prone to being affected. Without this knowledge, current mRNA vaccines against COVID-19 must be prepared and shipped under intense refrigeration, and are unlikely to reach more than a tiny fraction of human beings on the planet unless they can be stabilized.\n\n---\n\n### Our Goal: Predict the degradation rates at various locations along RNA sequence\n\n---\n\n### What do we have?\n\nWe have a subset of an **Eterna dataset** comprising over **3000 RNA molecules** (which span a panoply of sequences and structures) and their **degradation rates** at each position.\n\nThere are multiple ground truth values provided in the training data. While the submission format requires all 5 to be predicted, only the following are scored: reactivity, deg_Mg_pH10, and deg_Mg_50C.\n\n#### Files\n* train.json - the training data\n* test.json - the test set, without any columns associated with the ground truth.\n* sample_submission.csv - a sample submission file in the correct format\n\n#### Columns\n\n* id - An arbitrary identifier for each sample.\n\n* seq_scored - (68 in Train and Public Test, 91 in Private Test) Integer value denoting the number of positions used in scoring with predicted values. This should match the length of reactivity, deg_* and *_error_* columns. Note that molecules used for the Private Test will be longer than those in the Train and Public Test data, so the size of this vector will be different.\n\n* seq_length - (107 in Train and Public Test, 130 in Private Test) Integer values, denotes the length of sequence. Note that molecules used for the Private Test will be longer than those in the Train and Public Test data, so the size of this vector will be different.\n\n* sequence - (1x107 string in Train and Public Test, 130 in Private Test) Describes the RNA sequence, a combination of A, G, U, and C for each sample. Should be 107 characters long, and the first 68 bases should correspond to the 68 positions specified in seq_scored (note: indexed starting at 0).\n\n* structure - (1x107 string in Train and Public Test, 130 in Private Test) An array of (, ), and . characters that describe whether a base is estimated to be paired or unpaired. Paired bases are denoted by opening and closing parentheses e.g. (....) means that base 0 is paired to base 5, and bases 1-4 are unpaired.\n\n* reactivity - (1x68 vector in Train and Public Test, 1x91 in Private Test) An array of floating point numbers, should have the same length as seq_scored. These numbers are * reactivity values for the first 68 bases as denoted in sequence, and used to determine the likely secondary structure of the RNA sample.\n\n* deg_pH10 - (1x68 vector in Train and Public Test, 1x91 in Private Test) An array of floating point numbers, should have the same length as seq_scored. These numbers are reactivity values for the first 68 bases as denoted in sequence, and used to determine the likelihood of degradation at the base/linkage after incubating without magnesium at high pH (pH 10).\n\n* deg_Mg_pH10 - (1x68 vector in Train and Public Test, 1x91 in Private Test) An array of floating point numbers, should have the same length as seq_scored. These numbers are reactivity values for the first 68 bases as denoted in sequence, and used to determine the likelihood of degradation at the base/linkage after incubating with magnesium in high pH (pH 10).\n\n* deg_50C - (1x68 vector in Train and Public Test, 1x91 in Private Test) An array of floating point numbers, should have the same length as seq_scored. These numbers are reactivity values for the first 68 bases as denoted in sequence, and used to determine the likelihood of degradation at the base/linkage after incubating without magnesium at high temperature (50 degrees Celsius).\n\n* deg_Mg_50C - (1x68 vector in Train and Public Test, 1x91 in Private Test) An array of floating point numbers, should have the same length as seq_scored. These numbers are reactivity values for the first 68 bases as denoted in sequence, and used to determine the likelihood of degradation at the base/linkage after incubating with magnesium at high temperature (50 degrees Celsius).\n\n* *_error_* - An array of floating point numbers, should have the same length as the corresponding reactivity or deg_* columns, calculated errors in experimental values obtained in reactivity and deg_* columns.\n\n* predicted_loop_type - (1x107 string) Describes the structural context (also referred to as 'loop type')of each character in sequence. Loop types assigned by bpRNA from Vienna RNAfold 2 structure. From the bpRNA_documentation: S: paired \"Stem\" M: Multiloop I: Internal loop B: Bulge H: Hairpin loop E: dangling End X: eXternal loop\n\n___\n\n[*Competition Page*](https://www.kaggle.com/c/stanford-covid-vaccine/overview)"},{"metadata":{},"cell_type":"markdown","source":"## 1. Import Necessary Libraries"},{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true,"_kg_hide-input":true},"cell_type":"code","source":"import pandas as pd\nimport os\nimport numpy as np\n\nimport matplotlib.pyplot as plt\nimport seaborn as sns\n\nplt.style.use('fivethirtyeight')\n\nSEED = 2020\n\ndef seed_everything(SEED):\n    np.random.seed(SEED)\n#     tf.random.set_seed(SEED)\n    os.environ['PYTHONHASHSEED'] = str(SEED)\n\nseed_everything(SEED)","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","collapsed":true,"_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":false},"cell_type":"markdown","source":"## 2. Load Data"},{"metadata":{"trusted":true},"cell_type":"code","source":"PATH = '/kaggle/input/stanford-covid-vaccine'\nos.listdir(PATH)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"train = pd.read_json(os.path.join(PATH,'train.json'),lines=True).drop('index',axis=1)\ntest = pd.read_json(os.path.join(PATH,'test.json'),lines=True).drop('index',axis=1)\nsub = pd.read_csv(os.path.join(PATH,'sample_submission.csv'))","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## 3. Exploratory Data-Analysis"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"print(\"Train data shape: \",train.shape)\nprint(\"Test data shape: \",test.shape)\nprint(\"Sample submission shape: \",sub.shape)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**Why we have different size than test data for sample submission file?**  \n* Test data has **629** samples for **Public Test data** & new **3005** samples for **Private Test data**. (Total 629+3005 = 3634)\n* Public test data has **Sequence length of 107**. Hence, for public test data there will be total **(629*107) = 67,303 rows**\n* Private test data has **Sequence length of 130**. Hence, for private test data there will be total **(3005*130) = 3,90,650 rows**\n\n* Hence, Total rows in submission = 67,303 + 3,90,650 = **4,57,953**"},{"metadata":{"_kg_hide-input":false,"trusted":true},"cell_type":"code","source":"train.head()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"test.head()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### How the test data was produced?\n\nPublic test has **629** samples. This **629** samples were choosen from base data of **3029** in a way that it satisfies below 3 conditions:\n1. Minimum value across all 5 conditions must be greater than -0.5.\n2. Mean signal/noise across all 5 conditions must be greater than 1.0. [Signal/noise is defined as mean( measurement value over 68 nts )/mean( statistical error in measurement value over 68 nts)]\n3. To help ensure sequence diversity, the resulting sequences were clustered into clusters with less than 50% sequence similarity, and the 629 test set sequences were chosen from clusters with 3 or fewer members. That is, any sequence in the test set should be sequence similar to at most 2 other sequences.\n\n---\nIt was filtered based on three criteria detailed above to ensure that this subset is not dominated by any large cluster of RNA molecules with poor data, which might bias the public leaderboard. The remaining **2400 sequences** for which we have data are in **train.json**.\n\nNote that these filters have not been applied to the 2400 RNAs in the public training data train.json — some of those measurements have negative values or poor signal-to-noise, or some RNA sequences have near-identical sequences in that set. But we are providing all those data in case competitors can squeeze out more signal.\n\nThe three filters noted above will also **not be applied to Private Test on 3005 sequences.**"},{"metadata":{"trusted":true},"cell_type":"code","source":"train.columns","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**What are these signal_to_noise & SN_filter columns?**\n* **signal_to_noise** is defined as **mean( measurement value over 68 nts )/mean( statistical error in measurement value over 68 nts)**\n  \n\n* **SN_filter** is **1** is it satisfy the below given conditions or else it is **0**\n    \n    1. Minimum value across all 5 conditions must be greater than -0.5.\n    2. Mean signal/noise across all 5 conditions must be greater than 1.0.\n    \nHere, 5 conditions are the **ground-truth** values."},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"fig, ax = plt.subplots(1,2,figsize=(20,5))\n\ntrain['signal_to_noise'].plot.kde(ax=ax[0])\nax[0].set_title('Signal/Noise')\n\nsns.countplot(data=train,y='SN_filter',ax=ax[1])\nax[1].set_title('SN_filter')\n\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"plt.figure(figsize=(20,2))\nsns.boxplot(data=train,x='signal_to_noise')\n\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* We also have **negative** values of signal/noise\n* We can also see samples with **very high** signal/noise values *(could be Potential Outliers)*"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"print(\"Number of samples with -ev signal/noise values: \",train[train['signal_to_noise']<0].shape[0])\n\nQ1 = np.percentile(train['signal_to_noise'],q=25)\nQ3 = np.percentile(train['signal_to_noise'],q=75)\nIQR = Q3 - Q1\nprint(\"Number of samples with too high signal/noise values\", train[train['signal_to_noise']>Q3+1.5*IQR].shape[0])","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"train.seq_length.value_counts()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* All the samples in the train data are of **107** sequence length"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"test.seq_length.value_counts()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* Test data has both **107** & **130** sequence length samples"},{"metadata":{},"cell_type":"markdown","source":"### Targets Averaged over Positions"},{"metadata":{"trusted":true},"cell_type":"code","source":"avg_reactivity = np.array(list(map(np.array,train.reactivity))).mean(axis=0)\navg_deg_50C = np.array(list(map(np.array,train.deg_50C))).mean(axis=0)\navg_deg_pH10 = np.array(list(map(np.array,train.deg_pH10))).mean(axis=0)\n\navg_deg_Mg_50C = np.array(list(map(np.array,train.deg_Mg_50C))).mean(axis=0)\navg_deg_Mg_pH10 = np.array(list(map(np.array,train.deg_Mg_pH10))).mean(axis=0)","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"fig, ax = plt.subplots(1,3,figsize=(20,4))\n\n# Distribution of Reactivity Averaged over position\nsns.kdeplot(avg_reactivity,ax=ax[0])\nax[0].set_title('Distribution of Reactivity Averaged over position',size=15)\n\n# Distribution of deg_50C Averaged over position\nsns.kdeplot(avg_deg_50C,ax=ax[1])\nax[1].set_title('Distribution of deg_50C Averaged over position',size=15)\n\n# Distribution of deg_pH10 Averaged over position\nsns.kdeplot(avg_deg_pH10,ax=ax[2])\nax[2].set_title('Distribution of deg_pH10 Averaged over position',size=15)\n\n\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"plt.figure(figsize=(20,8))\n\nsns.lineplot(x=range(68),y=avg_reactivity,label='reactivity')\nsns.lineplot(x=range(68),y=avg_deg_50C,label='deg_50C')\nsns.lineplot(x=range(68),y=avg_deg_pH10,label='deg_ph10')\n\nplt.xlabel('Positions')\nplt.xticks(range(0,68,2))\nplt.ylabel('Values')\nplt.title('Average Target Values (w/o Mg) V/S Positions')\n\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* We can see high values of **Degradation & Reactivity** at the beginning of the sequence\n* There is high degradation because of **pH10** at the beginning but no such pattern for rest of the positions.\n* We can see some correlation between **deg_50C & deg_pH10**"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"plt.figure(figsize=(20,8))\nsns.regplot(x=avg_deg_50C,y=avg_deg_pH10)\n\nplt.title('Average deg_50C V/S deg_pH10')\nplt.ylabel('deg_50C')\nplt.xlabel('deg_pH10')\n\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"print(\"Correlation Coeff between avg_deg_50C & avg_deg_pH10: \",np.corrcoef(avg_deg_50C,avg_deg_pH10)[0][1])","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"fig, ax = plt.subplots(1,2,figsize=(20,4))\n\n# Distribution of deg_50C Averaged over position\nsns.kdeplot(avg_deg_Mg_50C,ax=ax[0])\nax[0].set_title('Distribution of deg_Mg_50C Averaged over position',size=15)\n\n\n# Distribution of deg_pH10 Averaged over position\nsns.kdeplot(avg_deg_Mg_pH10,ax=ax[1])\nax[1].set_title('Distribution of deg_Mg_pH10 Averaged over position',size=15)\n\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"plt.figure(figsize=(20,8))\n\nsns.lineplot(x=range(68),y=avg_deg_Mg_50C,label='deg_Mg_50C')\nsns.lineplot(x=range(68),y=avg_deg_Mg_pH10,label='deg_Mg_pH10')\n\nplt.xlabel('Positions')\nplt.xticks(range(0,68,2))\nplt.ylabel('Values')\nplt.title('Average Target Values (w Mg) V/S Positions')\n\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* Most of the time **deg_Mg_pH10** is more than **deg_Mg_50C**\n* And we can see some **correlation** between them"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"print(\"Correlation Coeff between avg_deg_Mg_50C & avg_deg_Mg_pH10: \",np.corrcoef(avg_deg_Mg_50C,avg_deg_Mg_pH10)[0][1])","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* **deg_Mg_50C & deg_Mg_pH10** are highly correlated"},{"metadata":{},"cell_type":"markdown","source":"### Distribution of Target at a Random Position"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"pos = np.random.choice(68)\n\nfig, ax = plt.subplots(1,3,figsize=(20,4))\n\n# Distribution of Reactivity at Random position\nsns.kdeplot(np.array(list(map(np.array,train.reactivity)))[:,pos],ax=ax[0])\nax[0].set_title(f'Distribution of Reactivity at position-{pos}',size=15)\n\n# Distribution of deg_50C at Random position\nsns.kdeplot(np.array(list(map(np.array,train.deg_50C)))[:,pos],ax=ax[1])\nax[1].set_title(f'Distribution of deg_50C at position-{pos}',size=15)\n\n# Distribution of deg_pH10 at Random position\nsns.kdeplot(np.array(list(map(np.array,train.deg_pH10)))[:,pos],ax=ax[2])\nax[2].set_title(f'Distribution of deg_pH10 at position-{pos}',size=15)\n\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"fig, ax = plt.subplots(1,2,figsize=(20,4))\n\n# Distribution of deg_50C at Random position\nsns.kdeplot(np.array(list(map(np.array,train.deg_Mg_50C)))[:,pos],ax=ax[0])\nax[0].set_title(f'Distribution of deg_Mg_50C at position-{pos}',size=15)\n\n\n# Distribution of deg_pH10 at Random position\nsns.kdeplot(np.array(list(map(np.array,train.deg_Mg_pH10)))[:,pos],ax=ax[1])\nax[1].set_title(f'Distribution of deg_Mg_pH10 at position-{pos}',size=15)\n\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Errors in Calculation (Averaged over positions)"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"y = ['reactivity_error','deg_error_Mg_pH10', 'deg_error_pH10', 'deg_error_Mg_50C','deg_error_50C']\nx = [np.array(list(map(np.array,train[col]))).mean(axis=0) for col in y]\n\nplt.figure(figsize=(20,5))\n\nsns.boxplot(y=y,x=x)\n\nplt.xlabel('Error values')\nplt.title('Average Errors in Calculation of Targets')\n\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* Very high values of Errors in **deg_error_pH10 & deg_error_50C** but maybe we should not worry about these targets because evaluation is not based on these targets\n* Maybe this could be a reason why our models will not be scored on these two targets"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"plt.figure(figsize=(20,8))\n\nfor i in range(len(y)):\n    sns.lineplot(x=range(68),y=x[i],label=y[i])\n    \nplt.xlabel('Positions')\nplt.xticks(range(0,68,2))\nplt.ylabel('Error')\nplt.title('Error V/S Position')\n\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* We can see an **interesting pattern** here. We have high values of **errors** between postions **0 and 4**, but then it starts **decreasing constantly** as we go ahead\n* Now let's analyse if all samples are contributing to these high values or there are only few smaples which are driving these values"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"pos = np.random.choice(68)\n\ny = ['reactivity_error','deg_error_Mg_pH10', 'deg_error_pH10', 'deg_error_Mg_50C','deg_error_50C']\nx = [np.array(list(map(np.array,train[col])))[:,pos] for col in y]\n\nplt.figure(figsize=(20,5))\nplt.title(f'Error Distribution at position - {pos}')\nplt.xlabel('Error')\n\nsns.boxplot(y=y,x=x)\n\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* We can see there are few samples which are having high values of errors\n* Now let's try to filter these samples\n* We can try using **SN_filter** for this"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"y = ['reactivity_error','deg_error_Mg_pH10', 'deg_error_pH10', 'deg_error_Mg_50C','deg_error_50C']\nx = [np.array(list(map(np.array,train[train.SN_filter==1][col])))[:,pos] for col in y]\n\nplt.figure(figsize=(20,5))\nplt.title(f'(Filtered) Error Distribution at position - {pos}')\nplt.xlabel('Error')\n\nsns.boxplot(y=y,x=x)\n\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* Samples with **SN_filter = 1** are having low values of errors\n* We can use this **filter** to train our model only on these samples but as the **private test** data is **not filtered**, it might only give good results on **Public LB** but not on **Private LB**"},{"metadata":{},"cell_type":"markdown","source":"### RNA Structure"},{"metadata":{"_kg_hide-input":true,"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"!pip3 install -q forgi[all]\n!conda install -y -c bioconda viennarna","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import forgi.graph.bulge_graph as fgb\nimport forgi.visual.mplotlib as fvm","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"def plot_sample(sample):\n    \n    \"\"\"Source: https://www.kaggle.com/erelin6613/openvaccine-rna-visualization\"\"\"\n    \n    struct = sample['structure']\n    seq = sample['sequence']\n    bg = fgb.BulgeGraph.from_fasta_text(f'>rna1\\n{struct}\\n{seq}')[0]\n    \n    plt.figure(figsize=(20,8))\n    fvm.plot_rna(bg)\n    plt.title(f\"RNA Structure (id: {sample.id})\")\n    plt.show()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"sample = train.iloc[np.random.choice(train.shape[0])]\nplot_sample(sample)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"print(\"Predicted Loop type: \",sample['predicted_loop_type'])","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* Here, we can see in image there is a **loop-m1** but for the this same loop, we have different **predicted loop type** in our data -> **loop type - X**\n* Why it is like this? "},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"from collections import defaultdict\n\nreactivity = defaultdict(lambda: [])\ndeg_Mg_50C = defaultdict(lambda: [])\ndeg_Mg_pH10 = defaultdict(lambda: [])\n\nfor i in range(len(sample['reactivity'])):\n    reactivity[sample['structure'][i]].append(float(sample['reactivity'][i]))\n    deg_Mg_50C[sample['structure'][i]].append(float(sample['deg_Mg_50C'][i]))\n    deg_Mg_pH10[sample['structure'][i]].append(float(sample['deg_Mg_pH10'][i]))\n\nplt.figure(figsize=(20,5))\nfor key in reactivity.keys():\n    sns.kdeplot(data=reactivity[key],label=key)\n\nplt.title('Structure wise Distribution of Reactivity of a Random Sample')\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"plt.figure(figsize=(20,5))\nfor key in reactivity.keys():\n    sns.kdeplot(data=deg_Mg_50C[key],label=key)\n\nplt.title('Structure wise Distribution of deg_Mg_50C of a Random Sample')\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"plt.figure(figsize=(20,5))\nfor key in reactivity.keys():\n    sns.kdeplot(data=deg_Mg_pH10[key],label=key)\n\nplt.title('Structure wise Distribution of deg_Mg_pH10 of a Random Sample')\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Pair wise Exploration of Degradation"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"reactivityDict = defaultdict(lambda: [])\n\nfor index in range(train.shape[0]):\n    \n    sample = train.iloc[index]\n\n    structure = sample['structure']\n    sequence = sample['sequence']\n    reactivity = sample['reactivity']\n\n    q = []\n\n    for i,s in enumerate(structure[:len(reactivity)]):\n        if s=='.':\n            reactivityDict[sequence[i]].append(reactivity[i])\n        elif s=='(':\n            q.append(i)\n        elif s==')':\n            j = q.pop(0)\n            key = \"-\".join(sorted([sequence[i],sequence[j]]))\n            reactivityDict[key].append(reactivity[i])\n            reactivityDict[key].append(reactivity[j])","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### Base-Pair wise Distribution of Reactivity"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"fig, ax = plt.subplots(len(reactivityDict.keys()),1,figsize=(20,2*len(reactivityDict.keys())),sharex=True)\n\nfor i, key in enumerate(reactivityDict.keys()):\n    sns.boxplot(x=reactivityDict[key],ax=ax[i])\n    ax[i].set_ylabel(key)\n\nplt.xlabel('Reactivity')\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* Single bases are having bigger range of **Reactivity** than **Paired Bases**.\n* Most of the pairs have distribution around **0**"}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}